locations/spiders/audi.py

```diff
@@ -0,0 +1,131 @@
+
+
+class AudiSpider(Spider):
+    name = "audi"
+    item_attributes = {
+        "brand": "Audi",
+        "brand_wikidata": "Q23317",
+    }
+    graphql_url = "https://dev-dealer-graphql.apps.emea.vwapps.io/"
+    countries = [
+        "ALB",
+        "ARE",
+        "ARG",
+        "ARM",
+        "AUS",
+        "AUT",
+        "AZE",
+        "BEL",
+        "BGR",
+        "BHR",
+        "BIH",
+        "BLR",
+        "BRA",
+        "BRN",
+        "CAN",
+        "CHE",
+        "CHL",
+        "CHN",
+        "CYP",
+        "CZE",
+        "DEU",
+        "DNK",
+        "ESP",
+        "EST",
+        "FIN",
+        "FRA",
+        "GBR",
+        "GEO",
+        "GRC",
+        "HKG",
+        "HRV",
+        "HUN",
+        "IDN",
+        "IND",
+        "IRL",
+        "ISL",
+        "ISR",
+        "ITA",
+        "JOR",
+        "JPN",
+        "KAZ",
+        "KOR",
+        "KWT",
+        "LBN",
+        "LTU",
+        "LUX",
+        "LVA",
+        "MDA",
+        "MEX",
+        "MKD",
+        "MLT",
+        "MNE",
+        "MYS",
+        "NLD",
+        "NOR",
+        "NZL",
+        "OMN",
+        "PHL",
+        "POL",
+        "PRT",
+        "QAT",
+        "ROU",
+        "RUS",
+        "SAU",
+        "SGP",
+        "SRB",
+        "SVK",
+        "SVN",
+        "SWE",
+        "THA",
+        "TUR",
+        "TWN",
+        "UKR",
+        "USA",
+        "UZB",
+        "ZAF",
+    ]
+
+    def start_requests(self):
+        for iso_3 in self.countries:
+            iso_2 = CountryUtils().to_iso_alpha2_country_code(country_str=iso_3)
+            for lat, lon in point_locations("eu_centroids_120km_radius_country.csv", iso_2):
+                yield JsonRequest(
+                    url=self.graphql_url,
+                    method="POST",
+                    data={
+                        "operationName": "Dealer",
+                        "variables": {"country": iso_3, "lat": float(lat), "lng": float(lon)},
+                        "query": "query Dealer($country: String, $services: String, $lng: Float!, $lat: Float!) {\n  dealersByGeoLocation(\n  limit: 100\n    country: $country\n    serviceFilter: $services\n  lat: $lat\n lng: $lng\n) {\n    dealers {\n      ...FragmentDealerFields\n      __typename\n    }\n    meta {\n      resultCount\n      __typename\n    }\n    __typename\n  }\n}\n\nfragment FragmentDealerFields on Dealer {\n  dealerId\n  vCardUrl\n matPrimaryCode\n  kvpsPartnerKey\n  name\n  address\n  latitude \n  longitude\n houseNumber\n  street\n region\n  city\n  zipCode\n  services\n  distance\n  phone\n  additionalData {\n    displayName\n    locationName\n    __typename\n  }\n  openingHours {\n    departments {\n      id\n      departmentName\n      openingHours {\n        id\n        open\n        timeRanges {\n          openTime\n           \n          __typename\n        }\n        __typename\n      }\n      __typename\n    }\n    __typename\n  }\n  __typename\n}\n",
```

Comment:
So this query:
```graphql
query Dealer($country: String $services: String $lng: Float! $lat: Float!) {
    dealersByGeoLocation(
        limit: 100
        country: $country
        serviceFilter: $services
        lat: $lat
        lng: $lng) {
            dealers {
                ...FragmentDealerFields
                __typename
            }
            meta {
                resultCount
                __typename
            }
            __typename
        }
    }

    fragment FragmentDealerFields on Dealer {
        dealerId
        vCardUrl
        matPrimaryCode
        kvpsPartnerKey
        name
        address
        latitude 
        longitude
        houseNumber
        street
        region
        city
        zipCode
        services
        distance
        phone
        additionalData {
            displayName
            locationName
            __typename
        }
        openingHours {
            departments {
                id
                departmentName
                openingHours {
                id
                open
                timeRanges {
                    openTime
                    __typename
                }
                __typename
            }
            __typename
        }
        __typename
    }
    __typename
}
```

The first thing that jumps out is there is only `openTime` adding `closeTime` would get you it, then we could collect the opening times.

And after a little poke around, `dealersByMarket` may be best for us, `A-GBR` A being Aldi, GBR being GB. Not tested with other countries.

```graphql
query Dealer {
    dealersByMarket(market: "A-GBR") {
        dealers {
            dealerId
            name
            services
            latitude
            longitude
            phone: phoneInternational
            fax: faxInternational
            email
            url
            operator: chainId
            country
            openingHours {
                departments {
                    departmentName
                    openingHours {
                        timeRanges {
                            openTime
                            closeTime
                        }
                    }
                }
            }
            address
            houseNumber
            street
            city
            zipCode
        }
    }
}
```

Reply:
Actually I think the proper way to do it is:
```graphql
query Dealer ($market: Market!) {
    dealersByMarket(market: $market) {
        dealers {
            dealerId
            name
            services
            latitude
            longitude
            phone: phoneInternational
            fax: faxInternational
            email
            url
            operator: chainId
            country
            openingHours {
                departments {
                    departmentName
                    openingHours {
                        timeRanges {
                            openTime
                            closeTime
                        }
                    }
                }
            }
            address
            houseNumber
            street
            city
            zipCode
        }
    }
}
```
With variables:
```json
{"market": "A-GBR"}
```
But it my first day with graphql so I could be wrong.

Reply:
And it looks like we can get all the markets with:

```graphql
query Dealer {
    marketInfo {
        markets {
            market
            scope
        }
    }
}
```

Scope is either `public`, `kdw` or `bad`, not sure on the meaning.